00X-tracing.md

```diff
@@ -0,0 +1,579 @@
+   using an external native module, there may be some additional optimizations that can be done
+   only through integration with the GC and/or JIT.
+
+### 1.2 Background
+The [Node.js Diagnostics WG](https://github.com/nodejs/diagnostics) became aware of a
+high-performance tracing framework being integrated from Chromium into v8. They agreed that the
+Node.js engine could also benefit from using the same tracing framework. See the
+[meeting notes from initial tracing discussion](https://github.com/nodejs/diagnostics/blob/master/wg-meetings/2016-06-01.md)
+and the [related WG issue](https://github.com/nodejs/diagnostics/issues/53).
+
+Integration of the C++ trace event framework into Node was
+[completed recently](https://github.com/nodejs/node/pull/9304). It can be
+[enabled at runtime via node command-line options](https://github.com/nodejs/node/blob/master/doc/api/tracing.md).
+
+Note the Node engine (C++ code) is not yet *instrumented* with calls to the tracing macros. That
+work is planned, but is outside the scope of this proposal.
+
+## 2. Tracing emitter APIs
+The tracing emitter APIs enable JavaScript code to emit tracing events of any of the supported
+types along with associated event metadata. The APIs are primarily intended to be used in the
+following scenarios:
+ * Instrumentation of application code, enabling application developers to collect performance
+   and diagnostic tracing data about their application code.
+ * Instrumentation of library code, enabling developers of applications consuming the library
+   to collect performance and diagnostic tracing data about the library, either separate from or
+   combined with application and engine tracing.
+
+The tracing emitter APIs might also be used as a low-level output format for higher-level tracing
+or logging frameworks. For example, logging frameworks that enable pluggable "sinks" for log events
+may support Node.js tracing as another sink option. However, since these APIs are primarily
+designed for performance tracing, they are not optimized for logging scenarios. For example,
+they don't define any log-level metadata or filtering based on it.
+
+### 2.1 New core module
+The proposed new tracing emitter APIs (and listener APIs) belong in a new Node.js core module:
+```javascript
+const tracing = require('tracing');
+```
+The "tracing" module name is [already reserved](https://www.npmjs.com/package/tracing) for this
+purpose. Alternatively, "trace" might be a good name, however that would have to be reconciled
+against the existing ["trace" npm package that generates long stack traces using AsyncWrap](https://www.npmjs.com/package/trace).
+
+### 2.2 Example use
+```javascript
+// Enable recording for the 'example' category. (Can also enable via node command-line.)
+const category = 'example';
+tracing.enableRecording(category);
+
+tracing.emit(category, { eventType: 'instant', name: 'myEvent' });
```

Comment:
could `tracing` expose constants to pass in as eventType instead of expecting particular strings?